### PR TITLE
Feature: Showing Only Applied Parts Of Handgrading Rubric

### LINKED ISSUE #40: Add option to only show applied parts of handgrading rubric to students
Currently, students see the full handgrading rubric along with their handgrading scores. This would add an option that, when enabled, only shows the parts of the rubric that were applied to their submission.

If I'm understanding this correctly, this probably applies to:
- Annotations
- Checkboxes that deduct points

This would probably need to be implemented by restricting the data included in `HandgradingResult.handgrading_rubric`. ~~We'll need to consider what impact that filtering logic has on performance.~~

### PR DESCRIPTION
**Pull request by developStorm**
Resolves https://github.com/eecs-autograder/autograder.io/issues/40

### CODE REVIEW

**Comment by developStorm** [comment]
This is a PoC for this feature. I wanted to check if this looks good before moving forward with implementing the instructor toggle option for this.


image

I only applied "visible" annotation/checkbox, and this is how it looks like on student's end

image

**Comment by james-perretta** [comment]
Ah, the reason I mentioned performance in the issue description was because
I was conflating this with things that have to be computed for all groups'
final submissions at once, but obviously it only happens to one at a time,
so performance isn't an issue.

What you have looks good. There is still the consideration of whether it
makes sense to hide checkboxes that give positive points--I think that
warrants a discussion but wouldn't majorly change the implementation.

Thanks for the PR!

**Comment by developStorm** [comment]
Thanks for clarifying! I reviewed Gradescope's practice and looks like they are not hiding checkboxes:
image

However I'm pretty sure what @ekaprits wants is to hide every unapplied rubric from student, including those with positive points ; ) so I'll probably just do that first

Or making this behavior optional is also possible. I can see if that can be done without adding too much complexity.

**Comment by developStorm** [comment]
Alright! I've implemented both frontend and backend part for this feature. It involves this PR and these two PRs:

https://github.com/eecs-autograder/ag-website-vue/pull/500 
https://github.com/eecs-autograder/ag-client-typescript/pull/147

I've roughly tested it locally and it works without problems. Let me know what you think 😺

**Comment by james-perretta** [comment]
@developStorm I don't suppose you could add a few test cases for this?

File: autograder/handgrading/views/handgrading_result_views.py [hunk 1]
```diff
@@ -85,7 +89,34 @@ class HandgradingResultView(NestedModelView):
...
+        result: dict = group.handgrading_result.to_dict()
+
+        # Only filter the output if the user doesn't have privilege
+        # Can be changed to is_student, if the permission does not inherit (i.e. if admin/staff are not considered student)
+        can_access_all_rubrics = (
+            P(HandgradingAllRubricReleased)
+            | P(is_admin)
+            | P(is_staff)
+            | P(is_handgrader)
+        )
```
Comment (james-perretta) [inline autograder/handgrading/views/handgrading_result_views.py]: I'd prefer not to use the permissions objects in this way, can we change this to something like this?
```python
course = group.project.course
show_only_applied_rubric = (
    group.project.handgrading_rubric.show_only_applied_rubric_to_students
    and not course.is_staff(request.user)  # Admins count as staff
    and not course.is_handgrader(request.user)
)
```

**Comment by james-perretta** [comment]
And following up on the checkbox hiding, I'm hesitant to have the default behavior be hiding unselected checkboxes regardless of the points strategy selected. Can you double check that that's truly the behavior that you need, rather than that only being the case for "start at max and subtract"?

File: autograder/handgrading/views/handgrading_result_views.py (hunk 1 above)
Comment (developStorm) [reply autograder/handgrading/views/handgrading_result_views.py]: Resolved in a7a93614449817ad8cc0da1aafcad2a34ce38832

**Comment by developStorm** [comment]
@james-perretta Thanks for the review, and tests are ready now! 

I double-confirmed with the professor, he intends to use this feature to minimize the rubric leakage (i.e. to prevent students share rubrics with other students from subsequent semesters). From this perspective, he believes hiding all unapplied rubrics is necessary. This is different from Gradescope's design - which is more about preventing students from the misconception that they are being docked points. I added a tooltip in settings page to emphasis the effect of this feature.

**Comment by james-perretta** [comment]
@developStorm Ok thanks, we can go with this for the time being. I'll update the issue with a note to expand on the implementation to allow for fine-grained control of which annotations and checkboxes to hide.

I see a couple of minor linter errors (I have yet to finish migrating my CI to github actions since Travis CI changed their free tier, sorry about that). ~~In the interest of time so that you can release grades ASAP, I'm going to make a local copy of your branch, make the fixes and run the tests, and then make a new PR to merge.~~

Edit: I'm going to merge this as-is as soon as the tests finish running and let future me fix the pycodestyle warnings.

**Review by james-perretta — APPROVED** [review]

**Comment by james-perretta** [comment]
Tests passing locally, merging with minor pycodestyle warnings.

**Comment by developStorm** [comment]
Thanks @james-perretta! I didn't notice there's a linter in-place so I formatted my code with black's default rules. Maybe we should consider adding linter info to the local env setup guide?

**Comment by james-perretta** [comment]
@james-perretta Yes, you're right, sorry about that! I can add that along with getting CI up and running again...

**Pull request merged**